### ISSUE TITLE
COD

### ISSUE THREAD

**Issue by Bijing1998** [Bijing1998/CODLSM#1]
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, StratifiedKFold
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.ensemble import AdaBoostClassifier
from lightgbm import LGBMClassifier
from sklearn.metrics import roc_curve, roc_auc_score
from keras.models import Sequential
from keras.layers import Dense, LSTM, Conv1D, MaxPooling1D, Flatten
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
import matplotlib.pyplot as plt
import pickle

# Load data
df = pd.read_csv(r'26级样本.csv')

# Data preprocessing
imputer = SimpleImputer(strategy='mean')
scaler = StandardScaler()

# Select feature and target columns
X = df.iloc[:, 0:12]
y = df.iloc[:, 12]

# Impute missing values and scale features
X_imputed = imputer.fit_transform(X)
X_scaled = scaler.fit_transform(X_imputed)

# Split dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.1, random_state=50)

# Reshape for LSTM and CNN
X_train_reshaped = X_train.reshape((X_train.shape[0], X_train.shape[1], 1))
X_test_reshaped = X_test.reshape((X_test.shape[0], X_test.shape[1], 1))

# Define a function to create the optimized LSTM model
def create_optimized_lstm_model():
    model = Sequential()
    model.add(LSTM(100, input_shape=(X_train_reshaped.shape[1], 1), return_sequences=True))
    model.add(LSTM(50, return_sequences=False))
    model.add(Dense(1, activation='sigmoid'))
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

# Define a function to create the optimized CNN model
def create_optimized_cnn_model():
    model = Sequential()
    model.add(Conv1D(128, kernel_size=3, activation='relu', input_shape=(X_train_reshaped.shape[1], 1)))
    model.add(MaxPooling1D(pool_size=2))
    model.add(Conv1D(64, kernel_size=3, activation='relu'))
    model.add(MaxPooling1D(pool_size=2))
    model.add(Flatten())
    model.add(Dense(32, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

# Update the base learners with the optimized models
base_learners = [
    ('LGBM', LGBMClassifier(n_estimators=150, max_depth=3, learning_rate=0.01, random_state=42)),
    ('AdaBoost', AdaBoostClassifier(n_estimators=150, random_state=42)),
    ('LSTM', KerasClassifier(build_fn=create_optimized_lstm_model, epochs=20, batch_size=32, verbose=0)),
    ('CNN', KerasClassifier(build_fn=create_optimized_cnn_model, epochs=20, batch_size=32, verbose=0)),
    ('SVM', SVC(probability=True)),
    ('MLP', MLPClassifier(max_iter=1000)),
]

# Prepare for plotting
plt.figure(figsize=(7, 7))
plt.gca().set_facecolor((0.9, 0.9, 0.9, 0.5))
mean_fpr = np.linspace(0, 1, 100)

# Prepare to store results
results = []
best_curves = {}  # 用于存储最佳曲线

# Define colors using RGB values normalized to [0, 1]
colors = {
    'LGBM': (38 / 255, 70 / 255, 83 / 255),
    'AdaBoost': (40 / 255, 114 / 255, 113 / 255),
    'LSTM': (138 / 255, 176 / 255, 125 / 255),
    'CNN': (233 / 255, 196 / 255, 107 / 255),
    'SVM': (243 / 255, 162 / 255, 97 / 255),
    'MLP': (230 / 255, 111 / 255, 81 / 255)
}

# Add models to a dictionary for iteration
models = dict(base_learners)

for name, model in models.items():
    tprs = []
    aucs_train = []
    aucs_test = []
    best_auc = 0
    best_fpr = None
    best_tpr = None

    # 10-fold cross-validation
    skf = StratifiedKFold(n_splits=10, shuffle=True, random_state=42)
    for train_index, test_index in skf.split(X_train, y_train):
        X_fold_train, X_fold_test = X_train[train_index], X_train[test_index]
        y_fold_train, y_fold_test = y_train.iloc[train_index], y_train.iloc[test_index]

        if name in ['LSTM', 'CNN']:
            X_fold_train_reshaped = X_fold_train.reshape((X_fold_train.shape[0], X_fold_train.shape[1], 1))
            X_fold_test_reshaped = X_fold_test.reshape((X_fold_test.shape[0], X_fold_test.shape[1], 1))
            model.fit(X_fold_train_reshaped, y_fold_train)
        else:
            model.fit(X_fold_train, y_fold_train)

        y_pred_proba = model.predict_proba(X_fold_test)[:, 1]

        # Calculate AUC for test set
        auc_test = roc_auc_score(y_fold_test, y_pred_proba)

        if auc_test > best_auc:
            best_auc = auc_test
            fpr, tpr, _ = roc_curve(y_fold_test, y_pred_proba)
            best_fpr = fpr
            best_tpr = tpr

    # Store results
    results.append({
        'Model': name,
        'Best Test AUC': best_auc
    })

    # Print results
    print(f'Model: {name}, Best Test AUC: {best_auc:.3f}')

    # Plot best ROC curve
    plt.plot(best_fpr, best_tpr, label=f'{name} (AUC = {best_auc:.3f})', linewidth=2.5, color=colors[name])

# Plot the diagonal line
plt.plot([0, 1], [0, 1], color='navy', linestyle='--', linewidth=2)
ax = plt.gca()
ax.spines['top'].set_linewidth(2)    # 上边框粗细
ax.spines['bottom'].set_linewidth(2) # 下边框粗细
ax.spines['left'].set_linewidth(2)   # 左边框粗细
ax.spines['right'].set_linewidth(2)  # 右边框粗细
plt.xlim([0.0, 1.0])
plt.ylim([0.0, 1.05])
plt.tick_params(axis='both', which='major', labelsize=16)
plt.grid(linestyle=':', color='black', alpha=0.7)
plt.xlabel('1-Specificity', fontsize=18)
plt.ylabel('Sensitivity', fontsize=18)
plt.legend(loc="lower right", fontsize=16)

plt.show()
